### PR TITLE
Add a test with a 3Mb ordinals transaction

### PR DESCRIPTION
**Pull request by sstone**
This tx highlighted a bug in how we encoded script size.

### CODE REVIEW

File: src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+        }
+        i = 1uL
+        val input = ByteArrayInput(output.toByteArray())
+        while (i <= 33554432UL) {
+            var j = BtcSerializer.varint(input)
+            assertEquals(i - 1uL, j)
+            j = BtcSerializer.varint(input)
+            assertEquals(i, j)
+            i *= 2uL
+        }
```
Comment (t-bast) [inline src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt:43]: Can you explain what you are testing here? I don't understand the `i - 1` and the `i *= 2uL`

File: src/commonMain/kotlin/fr/acinq/bitcoin/BtcSerializer.kt [hunk 2]
```diff
@@ -150,12 +150,12 @@ public abstract class BtcSerializer<T> {
         @JvmStatic
         public fun writeVarint(input: ULong, out: Output) {
             when {
-                input < 0xFDuL -> writeUInt8(input.toUByte(), out)
-                input < 65535uL -> {
+                input < 253uL -> writeUInt8(input.toUByte(), out)
+                input <= 65535uL -> {
```
[the remark this answers is not in the document]
Comment (sstone) [reply src/commonMain/kotlin/fr/acinq/bitcoin/BtcSerializer.kt:154]: Yes bitcoin is little endian and LN is big endian. With the proposed changes we match what bitcoin core does which is always a good thing :)

File: src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt (hunk 1 above)
Comment (sstone) [reply src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt:43]: It's a basic test copied from bitcoin core that checks that we can read back what we wrote, using inputs that are supposed to be the trickiest to handle (`2^n` and `2^n - 1`).  We also have many implicit tests (through reference test vectors etc...) but none that triggered this bug.

Comment (t-bast) [reply src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt:43]: This really needs a comment, a test that readers don't understand isn't a good thing :)

File: src/commonMain/kotlin/fr/acinq/bitcoin/BtcSerializer.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (t-bast) [reply src/commonMain/kotlin/fr/acinq/bitcoin/BtcSerializer.kt:154]: I find it a bit confusing that the implementation uses `<` for the first case but `<=` for the others to be honest...but if that matches the `bitcoind` implementation why not

File: src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+    }
```
Comment (t-bast) [inline src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt:44]: I find it much clearer that way, what do you think:

```suggestion
// test that we can read back what we wrote using values that are trickier to encode (2^n - 1, 2^n, 2^n + 1)
        for (i in 0..31) {
            val output = ByteArrayOutput()
            val v = (1 shl i).toULong()
            BtcSerializer.writeVarint(v - 1UL, output)
            BtcSerializer.writeVarint(v, output)
            BtcSerializer.writeVarint(v + 1UL, output)
            val input = ByteArrayInput(output.toByteArray())
            assertEquals(BtcSerializer.varint(input), v - 1UL)
            assertEquals(BtcSerializer.varint(input), v)
            assertEquals(BtcSerializer.varint(input), v + 1UL)
        }
```

**Review by t-bast — APPROVED** [review]
LGTM!

**Review by remyers — CHANGES_REQUESTED** [review]
I compared to the core code and tested with and without the fixes to double check the tests catch any issues. I only had a question and found an issue with the test.

File: src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt [hunk 4]
```diff
@@ -0,0 +1,40 @@
...
+        assertEquals("fc", varintToHex(0xfcuL))
+        assertEquals("fdfd00", varintToHex(0xfduL))
+        assertEquals("fdffff", varintToHex(0xffffuL))
+        assertEquals("fe00000100", varintToHex(0x10000uL))
+        assertEquals("feffffffff", varintToHex(0xffffffffuL))
+        assertEquals("ff0000000001000000", varintToHex(0x100000000uL))
+        assertEquals("ffffffffffffffffff", varintToHex(0xffffffffffffffffuL))
+
+        // test that we can read back what we wrote using values that are trickier to encode (2^n - 1, 2^n, 2^n + 1)
+        for (i in 0..31) {
```
[the remark this answers is not in the document]
Comment (sstone) [reply src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt:28]: yes `compactSize` is more widely used than `varint`. You're right I will update the test.

**Review by t-bast — APPROVED** [review]

File: src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt (hunk 4 above)
[the remark this answers is not in the document]
Comment (sstone) [reply src/commonTest/kotlin/fr/acinq/bitcoin/BtcSerializerTestsCommon.kt]: done in https://github.com/ACINQ/bitcoin-kmp/pull/83/commits/6bfdb03164b36581392511c89a2f9108e88b529f

**Review by remyers — APPROVED** [review]
Looks good! 

small nit: 

`val v = (1uL shl i)`  uses lower case `uL`, a few lines down it's an uppercase `UL` and a few lines above it's `uL`. Both are valid, Kotlin docs say "uL and UL explicitly tag literal as unsigned long"

**Comment by sstone** [comment]
> small nit:
> 
> `val v = (1uL shl i)` uses lower case `uL`, a few lines down it's an uppercase `UL` and a few lines above it's `uL`. Both are valid, Kotlin docs say "uL and UL explicitly tag literal as unsigned long"

Ah now I have to fix :)

**Pull request merged**